### PR TITLE
Improved performance of rz_bv_copy_nbits and rz_bv_set_range

### LINKED ISSUE #4716: Enhance performance of `rz_bv_copy_nbits` and `rz_bv_set_range`
**Is your feature request related to a problem? Please describe.**

Performance could be better and we save energy which would make us better humans.

**Describe the solution you'd like**

Both functions do something like this:
```c
for (ut32 i = 0; i < nbit; ++i) {
   bool c = rz_bv_get(src, src_start_pos + i); // Not in case of rz_bv_set_range()
   rz_bv_set(dst, dst_start_pos + i, c);
}
```

At least if we deal with small bit vectors (`<= 64bit`) we could do this with shift, AND + OR.
This would save a bunch of calls and other shifts and ORs.

For larger bit vectors we could check if the range is large enough (e.g. 8 bits or more).
Then set the first bits until the rest of them is aligned to a multiple of a byte (for `RzBitVector->large_a`) and assign the rest of them via `memcopy`. 

**Describe alternatives you've considered**

None really.

**Additional context**

This would give us quite some computation back. In my use case this loop consumes around 2.3%.
I also think this is a required improvement before we merge resource intensive analysis algorithms based on it.

### PR DESCRIPTION
**Pull request by rajRishi22**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
This pull request optimizes the performance of the rz_bv_set_range function. The original implementation iterated through bits one at a time, leading to inefficiencies for large ranges. The updated implementation:

Processes aligned chunks of bits using system word size for faster operations.
Dynamically adjusts chunk size for different architectures (e.g., 32-bit, 64-bit).
Handles unaligned prefix and suffix bits separately while optimizing the main loop.
Adds robust boundary validation to ensure correctness.
This change reduces iteration overhead and improves performance while maintaining compatibility and correctness.

**Test plan**

1. Verify functionality for small ranges, large ranges, and edge cases (unaligned ranges).
2. Test on various architectures to confirm portability (32-bit, 64-bit).
3. Use unit tests to ensure the results match the original functionality.
4. Check for any regressions with existing test suites.

**Closing issues**
Partially addresses #4716 
...

### CODE REVIEW

File: librz/util/bitvector.c [hunk 1]
```diff
@@ -209,26 +210,63 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
  * \param nbit ut32, control the size of copy (in bits)
  * \return copied_size ut32, Actual copied size
  */
+
 RZ_API ut32 rz_bv_copy_nbits(RZ_NONNULL const RzBitVector *src, ut32 src_start_pos, RZ_NONNULL RzBitVector *dst, ut32 dst_start_pos, ut32 nbit) {
-	rz_return_val_if_fail(src && dst, 0);
+    rz_return_val_if_fail(src && dst, 0);
```
Comment (Rot127) [inline librz/util/bitvector.c]: Please run `clang-format-16` on those files. The indentation gets fixed this way.

File: librz/util/bitvector.c [hunk 2]
```diff
@@ -209,26 +210,63 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+        nbit -= chunk_size;
+        bits_copied += chunk_size;
+    }
+
+    // Handle remaining unaligned suffix bits
+    while (nbit > 0) {
+        bool bit = rz_bv_get(src, src_start_pos++);
+        rz_bv_set(dst, dst_start_pos++, bit);
+        --nbit;
+        ++bits_copied;
```
Comment (Rot127) [inline librz/util/bitvector.c]: No need to increment `bits_copied`. You can simply save the original `nbit` value and return it.
As I see it there is no other return value then `nbit` or `0`.

File: librz/util/bitvector.c [hunk 3]
```diff
@@ -4,6 +4,7 @@
 #include "rz_util.h"
 #include <stdlib.h>
 #include <stdio.h>
+#include <limits.h>
```
Comment (Rot127) [inline librz/util/bitvector.c]:
```suggestion
#include <rz_types.h>
```

`limits.h` should be included already there.

File: librz/util/bitvector.c [hunk 4]
```diff
@@ -209,26 +210,63 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
  */
+
 RZ_API ut32 rz_bv_copy_nbits(RZ_NONNULL const RzBitVector *src, ut32 src_start_pos, RZ_NONNULL RzBitVector *dst, ut32 dst_start_pos, ut32 nbit) {
-	rz_return_val_if_fail(src && dst, 0);
+    rz_return_val_if_fail(src && dst, 0);
 
-	ut32 max_nbit = RZ_MIN((src->len - src_start_pos),
-		(dst->len - dst_start_pos));
+    // Determine the chunk size (word size) dynamically
+    const ut32 chunk_size = sizeof(unsigned long) * CHAR_BIT; // Word size in bits
```
Comment (Rot127) [inline librz/util/bitvector.c]: This is fine for now. But we should eventually change it against `uint_fast8_t, uint_fast16_t, etc.`.
But these need to be added in `rz_base_types.h` as `typedef`.

File: librz/util/bitvector.c [hunk 5]
```diff
@@ -209,26 +210,63 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
 
-	// normal case here
-	for (ut32 i = 0; i < nbit; ++i) {
-		bool c = rz_bv_get(src, src_start_pos + i);
-		rz_bv_set(dst, dst_start_pos + i, c);
-	}
+    ut32 bits_copied = 0;
+
+    // Handle unaligned prefix
+    while ((src_start_pos % chunk_size != 0 || dst_start_pos % chunk_size != 0) && nbit > 0) {
```
Comment (Rot127) [inline librz/util/bitvector.c]: You can make `(src_start_pos % chunk_size != 0 || dst_start_pos % chunk_size != 0)` an `if` condition and then loop around `nbit > 0`.
This way you can return early within the `if` body.

File: librz/util/bitvector.c [hunk 6]
```diff
@@ -209,26 +210,63 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+
+    // Process aligned chunks
+    while (nbit >= chunk_size) {
+        // Get chunks from the source and destination
+        unsigned long src_chunk = rz_bv_get_chunk(src, src_start_pos / chunk_size);
+        unsigned long dst_chunk = rz_bv_get_chunk(dst, dst_start_pos / chunk_size);
 
-	return nbit;
+        // Create a mask for the bits to copy
+        unsigned long mask = (1UL << chunk_size) - 1;
```
Comment (Rot127) [inline librz/util/bitvector.c]: This is equivalent to `UINT32_MAX` (or whatever type we settle on later).

File: librz/util/bitvector.c (hunk 4 above)
Comment (Rot127) [reply librz/util/bitvector.c]: `sizeof()` is a macro. So you can move this to a constant.

File: librz/util/bitvector.c [hunk 7]
```diff
@@ -1480,18 +1518,43 @@ RZ_API ut64 rz_bv_to_ut64(RZ_NONNULL const RzBitVector *x) {
...
-	for (ut32 i = pos_start; i <= pos_end; ++i) {
-		rz_bv_set(bv, i, b);
-	}
+    if (pos_start > bv->len - 1 || pos_end > bv->len - 1 || pos_start > pos_end) {
+        return false;
+    }
 
-	return true;
+    // Determine the chunk size dynamically
+    const ut32 chunk_size = sizeof(unsigned long) * CHAR_BIT;
```
Comment (Rot127) [inline librz/util/bitvector.c]: Same as above. Please move it to a constant/`#define`.

**Review by Rot127 — CHANGES_REQUESTED** [review]
Nice start!
I can imagine we can even optimize the unaligned cases. But let's do this later.

I also changed your PR message. Because it closes https://github.com/rizinorg/rizin/issues/4716 only partially (missing unaligned cases).

You can add test cases in `test/unit/test_bitvector.c`.
Once everything is implemented and passes, we can run the Travis CI to test it on big endian machines.

File: librz/util/bitvector.c (hunk 4 above)
Comment (Rot127) [reply librz/util/bitvector.c]: cc @XVilka @thestr4ng3r Regarding the `fast` types.

Comment (XVilka) [reply librz/util/bitvector.c]: First we need to make sure all our supported systems have it - at least running on `dist-` branch is required.

Comment (Rot127) [reply librz/util/bitvector.c]: It is in the `C99` standard. But if they don't support it we could just default to a normal `uint` type.

Comment (rajRishi22) [reply librz/util/bitvector.c]: Hi @Rot127  and @XVilka  , first of all thanks a lot for such a detailed review of my code ,as a beginner  it was extremely helpful for me . I have tried to address all problems mentioned in this review.
Do i need to create rz_base_types.h for now , as i could not find that file ?

File: librz/util/bitvector.c [hunk 8]
```diff
@@ -209,26 +213,62 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
  */
+
 RZ_API ut32 rz_bv_copy_nbits(RZ_NONNULL const RzBitVector *src, ut32 src_start_pos, RZ_NONNULL RzBitVector *dst, ut32 dst_start_pos, ut32 nbit) {
-	rz_return_val_if_fail(src && dst, 0);
+    rz_return_val_if_fail(src && dst, 0);
 
-	ut32 max_nbit = RZ_MIN((src->len - src_start_pos),
-		(dst->len - dst_start_pos));
+    // Determine the chunk size (word size) dynamically
+    const ut32 chunk_size = SIZE_OF_UNSIGNED_LONG * CHAR_BIT; // Word size in bits
```
Comment (Rot127) [inline librz/util/bitvector.c]: You can move this whole expression into a `#define`. Macros allow some basic math

```
#define RZ_BV_CHUNK_SIZE (sizeof(unsigned long) * CHAR_BIT)
```

File: librz/util/bitvector.c [hunk 9]
```diff
@@ -209,26 +213,62 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+        unsigned long src_chunk = rz_bv_get_chunk(src, src_start_pos / chunk_size);
+        unsigned long dst_chunk = rz_bv_get_chunk(dst, dst_start_pos / chunk_size);
+
+        // Create a mask for the bits to copy
+		unsigned long mask = UINT32_MAX;
+        if (nbit < chunk_size) {
+            mask = (1UL << nbit) - 1;
+        }
+
+        // Merge chunks using the optimized approach
```
Comment (Rot127) [inline librz/util/bitvector.c]: Add the link here.

File: librz/util/bitvector.c [hunk 10]
```diff
@@ -209,26 +213,62 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+        unsigned long result = dst_chunk ^ ((dst_chunk ^ src_chunk) & mask);
+        rz_bv_set_chunk(dst, dst_start_pos / chunk_size, result);
+
+        src_start_pos += chunk_size;
+        dst_start_pos += chunk_size;
+        nbit -= chunk_size;
```
Comment (Rot127) [inline librz/util/bitvector.c]: `nbit` underflows here if `nbit < chunk_size`.

**Review by Rot127 — CHANGES_REQUESTED** [review]

File: librz/util/bitvector.c [hunk 11]
```diff
@@ -209,24 +211,60 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+		// Merge chunks using the optimized approach , reference : https://graphics.stanford.edu/~seander/bithacks.html#MaskedMerge
+		unsigned long result = dst_chunk ^ ((dst_chunk ^ src_chunk) & mask);
+		rz_bv_set_chunk(dst, dst_start_pos / RZ_BV_CHUNK_SIZE, result);
+
+		src_start_pos += RZ_BV_CHUNK_SIZE;
+		dst_start_pos += RZ_BV_CHUNK_SIZE;
+		if (nbit >= RV_BV_CHUNK_SIZE)
+			nbit -= RZ_BV_CHUNK_SIZE;
+		else
+			break;
```
Comment (wargio) [inline librz/util/bitvector.c]:
```suggestion
		if (nbit < RV_BV_CHUNK_SIZE) {
			break;
		}
		nbit -= RZ_BV_CHUNK_SIZE;
```

File: librz/util/bitvector.c [hunk 12]
```diff
@@ -209,24 +211,60 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+			rz_bv_set(dst, dst_start_pos++, bit);
+			--nbit;
+		}
+	}
+
+	// Process aligned chunks
+	while (nbit >= RZ_BV_CHUNK_SIZE) {
+		// Get chunks from the source and destination
+		unsigned long src_chunk = rz_bv_get_chunk(src, src_start_pos / RZ_BV_CHUNK_SIZE);
+		unsigned long dst_chunk = rz_bv_get_chunk(dst, dst_start_pos / RZ_BV_CHUNK_SIZE);
```
Comment (wargio) [inline librz/util/bitvector.c]: maybe i'm missing something, but this function seems not in the code. also please use `ut32` as type.

File: librz/util/bitvector.c [hunk 13]
```diff
@@ -209,24 +211,60 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+
+		// Create a mask for the bits to copy
+		unsigned long mask = UINT32_MAX;
```
Comment (wargio) [inline librz/util/bitvector.c]:
```suggestion
		ut32 mask = UT32_MAX;
```

File: librz/util/bitvector.c [hunk 14]
```diff
@@ -2,11 +2,14 @@
...
 
 #include "rz_util.h"
+#include <rz_types.h>
 #include <stdlib.h>
 #include <stdio.h>
 
 #define NELEM(N, ELEMPER) ((N + (ELEMPER)-1) / (ELEMPER))
 #define BV_ELEM_SIZE      8U
+#define RZ_BV_CHUNK_SIZE  (sizeof(unsigned long) * CHAR_BIT)
+#define SIZE_OF_UNSIGNED_LONG sizeof(unsigned long)
```
Comment (wargio) [inline librz/util/bitvector.c]: ut32 not unsigned long

**Review by wargio — CHANGES_REQUESTED** [review]
missing implementation of `rz_bv_get_chunk`

File: librz/util/bitvector.c [hunk 15]
```diff
@@ -202,6 +202,79 @@ RZ_API ut32 rz_bv_copy(RZ_NONNULL const RzBitVector *src, RZ_NONNULL RzBitVector
...
+		// The chunk is aligned to a 32-bit boundary
+		bv->data[word_idx] = chunk;
+	} else {
+		// The chunk spans across two 32-bit words
+		ut32 first_word = bv->data[word_idx];
+		ut32 second_word = bv->data[word_idx + 1];
+
+		// Clear the bits in the current chunk positions
+		first_word &= ~(0xFFFFFFFF >> bit_offset); // Clear the upper bits
+		second_word &= (0xFFFFFFFF >> (32 - bit_offset)); // Clear the lower bits
```
Comment (wargio) [inline librz/util/bitvector.c:266]:
```suggestion
		first_word &= ~(UT32_MAX >> bit_offset); // Clear the upper bits
		second_word &= (UT32_MAX >> (32 - bit_offset)); // Clear the lower bits
```

**Review by wargio — CHANGES_REQUESTED** [review]
You are touching something very critical within rizin, so please provide benchmark and C test for correctness and edge cases.

**Comment by Rot127** [comment]
Are you still interested in finishing this? If yes, please fix the build and add tests.
Otherwise, I would close it in the next days.

**Comment by rajRishi22** [comment]
Sorry , i had my college exams so i was busy , i will start working on this issue again from today.

**Comment by XVilka** [comment]
This code looks like AI-generated. Closing it.
